### ISSUE TITLE
Add support for View Transition API in Router

### ISSUE THREAD

**Issue by eneajaho** [angular/angular#49401]
### Which @angular/* package(s) are relevant/related to the feature request?

router

### Description

Adding support for View Transition API directly in the Router package may help simplify router animations that we already have.

### Proposed solution

View Transition API: https://developer.mozilla.org/en-US/docs/Web/API/View_Transitions_API

### Alternatives considered

Current route animations: https://angular.io/guide/route-animations

**Comment by atscott** [comment]
Yea, definitely. In the meantime, here's a fancy router outlet to play with (the router API does allow for custom router outlets): https://stackblitz.com/edit/angular-fcnbwy-7rkfyz?file=src%2Fapp%2Ffancy_router_outlet.ts. I wouldn't be surprised if it has some/many quirks. Use at your own risk 😉

**Comment by konstantindenerz** [comment]
Thank you @atscott  for the fancy router outlet ✨. 

I had tried it. I noticed something. I think that the RouterOutlet is not the best place to call startViewTransition in the current view transition browser implementation. Maybe you can help me there?

### StartViewTransition behaviour how I understand it:  
StartViewTransition creates a pseudo element tree with a static element (**old**), that represents e.g. the root element before transition (navigation) and target element (**new**) after transition.

### Use Case / Expectation
So if I'm navigating from a **list** route to a detail route in my demo application, then I think you only need a single startViewTransition invocation.

### Fancy Router Outlet
The fancy router outlet invokes startViewTransition two times in the case above. This leads to unexpected view transition behavior in my demo application.
I think you need to wrap both route deactivation and activation with just a single call.

### Change
Please take a look at this change:
https://github.com/konstantindenerz/angular/commit/8ea4d971801e2e144ad46c3d9514360f0abf1526#diff-03cb1ea13f0b5890eb24ce6fef7a42eed85429da0a90cdb48ab06b58c30cb1c1R741-R766

It's probably not the best place to call it and importing/using NgZone is undesirable, but for now my demo application works as expected.

I think the multiple calls to startViewTransition in the Fancy Router Outlet come from the idea of the enter/leave animation, but I can't imagine it with the view transitions.

**Comment by atscott** [comment]
@konstantindenerz Thanks for the input. Please note that the outlet implementation is absolutely not a proposal for how it should be implemented. However, it is in user-space and can be used while there is no support in the Angular Router.

**Comment by kbrilla** [comment]
Question: here in Chrome Developers youtube video on [morphing buttons ](https://www.youtube.com/watch?v=N2BKAKwGP6M) View Transition API is used not on routing but actually on dom changes and I was wandering are You thinking of way to use this api in that way in Angular? So to allow startViewTransition on changeDetection or for nextRender?

**Comment by JeanMeche** [comment]
@kbrilla This kind of transitions are handled by the animation module. Maybe in the future the animation module could use the new APIs.

**Comment by naveedahmed1** [comment]
FYI https://developer.chrome.com/blog/astro-view-transitions/

**Comment by sp90** [comment]
This PR #51314 will break anyone using v17 ssr

I made a small workaround to only include in the appConfig and not also in the server config

```ts
// app.config.ts
export const baseAppConfig: ApplicationConfig = {
  providers: [
    provideRouter(routes, withComponentInputBinding())
  ],
};

export const appConfig: ApplicationConfig = mergeApplicationConfig(baseAppConfig, {
  providers: [provideRouter(routes, withComponentInputBinding(), withViewTransitions())]
});

// app.config.server.ts
const serverConfig: ApplicationConfig = {
  providers: [provideServerRendering()]
};

export const config = mergeApplicationConfig(baseAppConfig, serverConfig);
``` 

But i would argue we need a better way to use it as of v17 release

**Comment by naveedahmed1** [comment]
It's a client side feature so I think it would make more sense to include it in client app config only right? this is what I did for Mustakbil.com.

We have SSR enabled and it seems to be working fine.

The only issue I faced is the flickering issue when we have a server rendered response available and then the client app takes control on client side and sync the server rendered view and applies the view transition on already painted view.

As a work around we have added a class to html element with name app-start and in css we have given it a transition name as a results the default fadein transition isn't applied.

And on the client side when the app initializes we remove that class from the html element. So that the view transitions start working for client side navigation.

But I believe it should be handled by the framework by default.

**Comment by sp90** [comment]
Im using angular v17 and their angular/ssr package and that breaks when adding this one.

Also because you're merging it into your `app.server.config.ts` 

So yes its a client behavior but if you're server is trying to evaluate it on a route it breaks ofc because document is not defined on the server

**Comment by JeanMeche** [comment]
@atscott Opened a PR to fix that issue #51814
